src/LibUvLoop.php

```diff
@@ -0,0 +1,263 @@
+    {
+        if (!isset($this->events[(int) $stream])) {
+            return;
+        }
+
+        unset($this->listeners[(int) $stream]['write']);
+
+        if (!isset($this->listeners[(int) $stream]['read'])
+            && !isset($this->listeners[(int) $stream]['write'])) {
+            unset($this->events[(int) $stream]);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function removeStream($stream)
+    {
+        if (isset($this->events[(int) $stream])) {
+
+            \uv_poll_stop($this->events[(int) $stream]);
+
+            unset($this->listeners[(int) $stream]['read']);
+            unset($this->listeners[(int) $stream]['write']);
+            unset($this->events[(int) $stream]);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addTimer($interval, callable $callback)
+    {
+        $timer = new Timer( $interval, $callback, false);
+
+        $callback = function () use ($timer) {
+            call_user_func($timer->getCallback(), $timer);
+
+            if ($this->isTimerActive($timer)) {
+                $this->cancelTimer($timer);
+            }
+        };
+
+        $event = \uv_timer_init($this->uv);
+        $this->timerEvents->attach($timer, $event);
+        \uv_timer_start(
+            $event,
+            $interval * 1000,
+            0,
+            $callback
+        );
+
+        return $timer;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addPeriodicTimer($interval, callable $callback)
+    {
+        $timer = new Timer($interval, $callback, true);
+
+        $callback = function () use ($timer) {
+            call_user_func($timer->getCallback(), $timer);
+        };
+
+        $event = \uv_timer_init($this->uv);
+        $this->timerEvents->attach($timer, $event);
+        \uv_timer_start(
+            $event,
+            $interval * 1000,
+            $interval * 1000,
+            $callback
+        );
+
+        return $timer;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function cancelTimer(TimerInterface $timer)
+    {
+        if (isset($this->timerEvents[$timer])) {
+            @\uv_timer_stop($this->timerEvents[$timer]);
+            $this->timerEvents->detach($timer);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isTimerActive(TimerInterface $timer)
+    {
+        return $this->timerEvents->contains($timer);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function futureTick(callable $listener)
+    {
+        $this->futureTickQueue->add($listener);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function run()
+    {
+        $this->running = true;
+
+        while ($this->running) {
+            $this->futureTickQueue->tick();
+
+            if ($this->futureTickQueue->isEmpty() && empty($this->events) && $this->timerEvents->count() === 0) {
+                break;
+            }
+
+            \uv_run($this->uv, \UV::RUN_NOWAIT);
```

Comment:
Always using `RUN_NOWAIT` will result in a busy loop.

Reply:
What do you suggest as an alternative?

Reply:
Depends on whether you have things scheduled for the next tick or not, see https://github.com/amphp/amp/blob/1cb074ac9078f1c89787b539d605083c5845ceab/lib/Loop/UvDriver.php#L174

Reply:
Addressed this, thanks :+1: